Add `issuanceLoadStat` to RA, exposed via debug server

ccppuu opened the pull request:
The `issuanceLoadStat` is a type backed by a circular `[]int` slice of size 60. Every time a certificate is issued the current minute is used as an index into this circular bucket slice and the corresponding int is incremented.

Simultaneously a Go routine monitors a ticker and every 60s updates an expvar integer to be the value of the past 5 minutes of bucket totals summed.

The count of successful issuances for the 5 minute window is exposed via the expvar debug server for the RA at port 8002 under the key "successfulIssuances".

Closes #1945

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,98 @@ type RegistrationAuthorityImpl struct {
+// else in NewRegistrationAuthority() then multiple invocations of the
+// constructor (e.g from unit tests) will panic with a "Reuse of exported var
+// name:" error from the expvar package.
+var issuanceExpvar *expvar.Int
```
Can this be `var issuanceExpvar = expvar.NewInt(...)` ?

ccppuu replied:
Yup! Not sure why I didn't do that to begin with. Pushed a commit to address.

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,93 @@ type RegistrationAuthorityImpl struct {
+	clk          clock.Clock
+}
+
+// Spawn a Go routine blocking on the ticker
```
s/Go routine/goroutine/

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,93 @@ type RegistrationAuthorityImpl struct {
+			count := i.countIssuances()
+			issuanceExpvar.Set(int64(count))
+			// Then block on the ticker until its time to update again
+			<-i.ticker
```
Did you consider a `time.Sleep` here? It might be simpler and we already have a mock Sleep in fakeclock.

jsha replied:
Actually nevermind - I think fakeclock's Sleep just adds to the stored time and returns immediately, which would make this fast-loop.

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,93 @@ type RegistrationAuthorityImpl struct {
+		}
+		sum += i.issuedCounts[index]
+	}
+	i.RUnlock()
```
Typically we put `defer i.RUnlock()` right after the lock is acquired so it's easy to see both are there.

Also, let's call the function receiver `ils` instead of `i` to avoid confusion with an index variable.

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,93 @@ type RegistrationAuthorityImpl struct {
+	// Increment the bucket for this minute's issuance count
+	cur := i.clk.Now().Minute()
+	i.Lock()
+	i.issuedCounts[cur] = i.issuedCounts[cur] + 1
```
I think you can do ++ here.

jsha commented on ra/ra.go:
```diff
@@ -72,6 +74,93 @@ type RegistrationAuthorityImpl struct {
+	i.Unlock()
+}
+
+func newIssuanceLoadStat(statName string, clk clock.Clock) *issuanceLoadStat {
```
Looks like statName isn't used?

ccppuu replied:
Good catch - leftover from before I realized the `expvar.NewInt` had to be global. I'll remove that param.

jsha commented:
I don't see any code that will zero out a bucket when we come back around to that minute. Thinking about it some, maybe getting that code right is too complicated, and instead we should have a constantly growing slice and always count the last N. Feels ugly, but a week is only 10k minutes, and even if we left Boulder running for a whole month it would only be 40k minutes. What do you think?

ccppuu commented:
After some chat offline I implemented a `lastBucket` int. In `issuedCert` when `cur != lastBucket` the bucket is zeroed before incrementing the count. I also added this behaviour to the test.

Ready for another review.

ccppuu commented:
Chat among the team has lead to a design change that is simpler, faster and more reliable. Revamping this PR.
